src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs

```diff
@@ -334,11 +344,27 @@ public async Task Invoke(HttpContext context)
                             // Save the image to the cache and send the response to the caller.
                             await this.cache.SetAsync(key, outStream, cachedImageMetadata);
 
-                            // Remove the resolver from the cache so we always resolve next request
+                            // Remove any resolver from the cache so we always resolve next request
                             // for the same key.
                             CacheResolverLru.TryRemove(key);
 
-                            await this.SendResponseAsync(imageContext, key, cachedImageMetadata, outStream, null);
```

Comment:
This was essentially the issue, where the response was only sent once.

Because this factory now produces a result that must be sent multiple times, I've disposed of the `outStream` at the end of this block, and am returning a response from the cache that it creates. Otherwise it needs a way to keep track of how many awaiters are using the same stream (and lock the stream, during each response), so it can then finally dispose it.

Maybe an area to improve, but this seemed like the simpler solution, and it only occurs during a write

Reply:
I'm still trying to understand the cause of our issue. We're only returning the metadata as the result of our Lazy so it should be safe to return the result multiple times since we're always requesting the stream from that metadata via a method. (we don't keep hold of the stream)

https://github.com/SixLabors/ImageSharp.Web/blob/8f67aae9d314a6fde96aae856027a655d6cbd2fa/src/ImageSharp.Web/Resolvers/PhysicalFileSystemResolver.cs#L33

Are we holding onto the result for too long?

Reply:
The issue is that in the current Lazy read worker is is only returning whether to refresh the disccache or not. 
It's actually doing the send of the image response inside the worker valuefactory.
And because the point of the lazy was to stop those valuefactories running multiple times, it was doing exactly that. Running it once. So 5 awaiters would await the result, but all the result would do is return or allow a refresh of the cached image. 

```
            (bool newOrUpdated, ImageMetadata sourceImageMetadata) =
                await this.IsNewOrUpdatedAsync(sourceImageResolver, imageContext, key); <-- the sendresponse happens once. no matter how many awaiters

            if (!newOrUpdated)
            {
                return;
            }
```

so the change is to get a cacheresolver back from the task.
then all awaiters can send the image response, each awaiter, opening the `CreateReadStream()`

Reply:
Ah ok.. So our numbers were junk so the loss avoidable. 

Just to clarify the corrected behavior.

Given A-C are unique URLs numbered 1-2

Request A (1) is received - This is **uncached** but is not blocked because there are no current workers processing that URL.
Request B (1) is received - This is **uncached** but is not blocked because there are no current workers processing that URL.
Request C (1) is received  - This is **cached** but is not blocked because there are no current workers processing that URL.
Request C (2) is received  - This is **cached** but is not blocked because C (1) already processing that URL is a cache read.
Request C (1) is complete.
Request C (2) is complete.
Request A (2) is received - This is **blocked** because A (1) is a write request and has not completed.
Request B (2) is received - This is **blocked** because B (1) is a write request and has not completed.
Request A (1) is complete.
Request A (2) is complete and reads from the newly cached file created by A (1).
Request B (1) is complete.
Request B (2) is complete and reads from the newly cached file created by B (1).

Reply:
> Just to clarify the corrected behavior.

Yes, that adds up to me.

The only comment I would make is just to clarify that the read workers still block/await
- Request C (1) is received - This is cached but is not blocked because there are no current workers processing that URL.
- Request C (2) is received - This is cached but is not blocked because C (1) already processing that URL is a cache read.

In this scenario C (1) is cached on disc, but not yet in LRU. C (1) blocks the LRU read (read of source metadata, and read of cache metadata). When it completes C (2) is also freed to return the same LRU result. (and when C (3) arrives later, it gets the LRU result, without doing the metadata reads).

I looked to see if the read worker could be removed (just to simplify), and the LRU Cache made so it's `GetOrAddAsync` could be blocked with the same technique, but because there is a source metadata read, and cache metadata read to do, which are in different LRU caches, it wasn't viable, and better to block the initial reads into the both of the LRU caches.

Reply:
Yep that is sensible and works for me. I think we’re good to go with this, nice job!